File: Call-Android/app/src/main/java/org/ar/call/BaseActivity.java
package org.ar.call;

import android.os.Bundle;

import com.kongzue.dialog.interfaces.OnDismissListener;
import com.kongzue.dialog.v3.MessageDialog;
import com.kongzue.dialog.v3.TipDialog;

import org.ar.call.utils.RTManager;
import org.ar.rtm.LocalInvitation;
import org.ar.rtm.RemoteInvitation;
import org.ar.rtm.RtmCallEventListener;
import org.ar.rtm.RtmClientListener;
import org.ar.rtm.RtmMessage;

import java.util.Map;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public  abstract class BaseActivity extends AppCompatActivity implements RtmClientListener, RtmCallEventListener {

    private boolean isReconnect = false;
    private MessageDialog reconnectDialog;
    private boolean needUnRegister = true;

    public void setNeedUnRegister(boolean needUnRegister) {
        this.needUnRegister = needUnRegister;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        needUnRegister = true;
        RTManager.INSTANCE.registerRtmEvent(this);
        RTManager.INSTANCE.registerCallListener(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        RTManager.INSTANCE.registerRtmEvent(this);
        RTManager.INSTANCE.registerCallListener(this);
    }


    @Override
    protected void onRestart() {
        super.onRestart();
        if ( RTManager.INSTANCE.getRemoteInvitation()!=null){
            //退到后台 如果有人呼叫 回到页面重新显示
            onRemoteInvitationReceived( RTManager.INSTANCE.getRemoteInvitation());
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (needUnRegister) {
            RTManager.INSTANCE.unRegisterRtmEvent(this);
            RTManager.INSTANCE.unRegisterCallListener(this);
        }
    }

    @Override
    public void onConnectionStateChanged(int state, int var2) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (state==4){
                    isReconnect = true;
                    reconnectDialog=MessageDialog.build(BaseActivity.this)
                            .setTitle("提示")
                            .setMessage("正在重连....")
                            .setOnOkButtonClickListener((baseDialog, v) -> true)
                            .setCancelable(false);
                    reconnectDialog.show();
                }else if (state ==3){
                    if (isReconnect){
                        if (reconnectDialog!=null){
                            reconnectDialog.doDismiss();
                        }
                        isReconnect= false;
                        TipDialog.show(BaseActivity.this, "重连成功！", TipDialog.TYPE.SUCCESS).setOnDismissListener(new OnDismissListener() {
                            @Override
                            public void onDismiss() {
                            }
                        });
                    }
                }
            }
        });
    }

    @Override
    public void onMessageReceived(RtmMessage var1, String var2) {

    }

    @Override
    public void onTokenExpired() {

    }

    @Override
    public void onPeersOnlineStatusChanged(Map<String, Integer> var1) {

    }

    @Override
    public void onLocalInvitationReceivedByPeer(LocalInvitation var1) {

    }

    @Override
    public void onLocalInvitationAccepted(LocalInvitation var1, String var2) {

    }

    @Override
    public void onLocalInvitationRefused(LocalInvitation var1, String var2) {

    }

    @Override
    public void onLocalInvitationCanceled(LocalInvitation var1) {

    }

    @Override
    public void onLocalInvitationFailure(LocalInvitation var1, int var2) {

    }

    @Override
    public void onRemoteInvitationReceived(RemoteInvitation var1) {

    }

    @Override
    public void onRemoteInvitationAccepted(RemoteInvitation var1) {

    }

    @Override
    public void onRemoteInvitationRefused(RemoteInvitation var1) {

    }

    @Override
    public void onRemoteInvitationCanceled(RemoteInvitation var1) {

    }

    @Override
    public void onRemoteInvitationFailure(RemoteInvitation var1, int var2) {

    }


}

File: Call-Android/app/src/main/java/org/ar/call/BaseActivity.kt
package org.ar.call

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.kongzue.dialog.interfaces.OnDismissListener
import com.kongzue.dialog.util.BaseDialog
import com.kongzue.dialog.v3.MessageDialog
import com.kongzue.dialog.v3.TipDialog
import org.ar.call.RtmManager.Companion.instance
import org.ar.rtm.*

abstract class BaseActivity : AppCompatActivity(), RtmClientListener, RtmCallEventListener {
    private var isReconnect = false
    private var reconnectDialog: MessageDialog? = null
    private var needUnRegister = true
    fun setNeedUnRegister(needUnRegister: Boolean) {
        this.needUnRegister = needUnRegister
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        needUnRegister = true
        instance.registerRtmEvent(this)
        instance.registerCallListener(this)
    }

    override fun onResume() {
        super.onResume()
        instance.registerRtmEvent(this)
        instance.registerCallListener(this)
    }

    override fun onStop() {
        super.onStop()
        if (needUnRegister) {
            instance.unRegisterRtmEvent(this)
            instance.unRegisterCallListener(this)
        }
    }

    override fun onConnectionStateChanged(state: Int, var2: Int) {
        runOnUiThread {
            if (state == 4) {
                isReconnect = true
                reconnectDialog = MessageDialog.build(this@BaseActivity)
                        .setTitle("提示")
                        .setMessage("正在重连....")
                        .setOnOkButtonClickListener { baseDialog: BaseDialog?, v: View? -> true }
                        .setCancelable(false)
                reconnectDialog?.show()
            } else if (state == 3) {
                if (isReconnect) {
                    reconnectDialog?.let {
                        it.doDismiss()
                    }
                    isReconnect = false
                    TipDialog.show(this@BaseActivity, "重连成功！", TipDialog.TYPE.SUCCESS).onDismissListener = OnDismissListener { }
                }
            }
        }
    }

    override fun onMessageReceived(var1: RtmMessage, var2: String) {}
    override fun onTokenExpired() {}
    override fun onPeersOnlineStatusChanged(var1: Map<String, Int>) {}
    override fun onLocalInvitationReceivedByPeer(var1: LocalInvitation) {}
    override fun onLocalInvitationAccepted(var1: LocalInvitation, var2: String) {}
    override fun onLocalInvitationRefused(var1: LocalInvitation, var2: String) {}
    override fun onLocalInvitationCanceled(var1: LocalInvitation) {}
    override fun onLocalInvitationFailure(var1: LocalInvitation, var2: Int) {}
    override fun onRemoteInvitationReceived(var1: RemoteInvitation) {}
    override fun onRemoteInvitationAccepted(var1: RemoteInvitation) {}
    override fun onRemoteInvitationRefused(var1: RemoteInvitation) {}
    override fun onRemoteInvitationCanceled(var1: RemoteInvitation) {}
    override fun onRemoteInvitationFailure(var1: RemoteInvitation, var2: Int) {}
}